lang/funcs/collection.go

```diff
@@ -841,11 +866,25 @@ var ValuesFunc = function.New(&function.Spec{
 	Params: []function.Parameter{
 		{
 			Name: "values",
-			Type: cty.Map(cty.DynamicPseudoType),
+			Type: cty.DynamicPseudoType,
 		},
 	},
 	Type: func(args []cty.Value) (ret cty.Type, err error) {
-		return cty.List(args[0].Type().ElementType()), nil
+		ty := args[0].Type()
+		if ty.IsMapType() {
+			return cty.List(ty.ElementType()), nil
+		} else if ty.IsObjectType() {
+			var tys []cty.Type
+			for _, v := range ty.AttributeTypes() {
+				tys = append(tys, v)
+			}
+			retType, _ := convert.UnifyUnsafe(tys)
+			if retType == cty.NilType {
+				return cty.NilType, fmt.Errorf("all arguments must have the same type")
+			}
+			return cty.List(retType), nil
```

Comment:
We can avoid the requirement that all arguments must have the same type here by returning a tuple type in this case:

```suggestion
			return cty.Tuple(tys), nil
```

I think this would also require a small tweak to each of the `return` statements in `Impl` below to handle the two possibilities:

```go
    if retVal.IsTupleType() {
        return cty.TupleVal(values), nil
    } else {
        if len(values) == 0 {
            return cty.ListValEmpty(retType.ElementType()), nil
        }
        return cty.ListVal(values), nil
    }
```

(we don't need to treat the empty list as special in the tuple case because "empty tuple" is a distinct type altogether, which is what'll be returned if an empty slice is passed to `cty.TupleVal`.)

Reply:
I'm slightly confused by the suggested change - should I return tuple type only in the case where the arguments don't have the same type, or should I drop the checking and just return a tuple type?

Reply:
I'd suggest the following:
* If the given value is of a map type, the return type is a list of the same element type.
* If the given value is of an object type, the return type is a tuple with the same element types as the given object type (in the order `objVal.ElementIterator` uses, which is always sorted.)

The tricky edge with always returning a tuple type is if the given value is an _unknown_ value of map type, since in that case it won't be possible to predict the corresponding result type because the number of elements in the map isn't known. This isn't an issue for unknown values of object type because the set of attribute names and types is part of the object type itself, not part of the value.

Reply:
Thanks, I think I got it this time!